### PR TITLE
Search by tags

### LINKED ISSUE #27: NuGet package is "Not compatible"
I tried to install the plugin `GitExtensions.SVN` via the PM. The plugin is available here: https://www.nuget.org/packages/GitExtensions.SVN/0.2.0.21

It has a dependency on `GitExtensions.Extensibility`, which is available on the same feed: https://www.nuget.org/packages/GitExtensions.Extensibility/0.1.0.31

**Reproduction steps:**

1. download and extract the portable build from https://github.com/gitextensions/gitextensions/pull/6664: https://ci.appveyor.com/api/buildjobs/d4jebf45vply353j/artifacts/Setup%2FGitExtensions-Portable-3.2.0.6454.zip
2. open GE and start the PluginManager plugin
3. search for `GitExtensions.SVN`
4. Nothing is found
image
5. However, the log states that the package `GitExtensions.SVN.0.2.0.21` was correctly found, but skipped for some unknown reason:
```
22/08/2019 22:08:24 root.Search(DEBUG)

Searching 'GitExtensions.SVN'.

22/08/2019 22:08:24 root.Search(DEBUG)

Loading page '0'.

22/08/2019 22:08:24 root.Search(DEBUG)

Searching in 'https://api.nuget.org/v3/index.json'.

22/08/2019 22:08:25 root.Search(DEBUG)

Found 'GitExtensions.SVN.0.2.0.21'.

22/08/2019 22:08:25 root.Search(DEBUG)

Package skipped.

22/08/2019 22:08:25 root.Search(DEBUG)

Loading page '1'.

22/08/2019 22:08:25 root.Search(DEBUG)

Search completed. Found '0' items.

```

**What I found so far**

I think I found a problem in https://github.com/gitextensions/gitextensions.pluginmanager/blob/3dad53b79112f75a6ae0b1c00801132e9a251bb0/src/PackageManager.NuGet/Services/NuGetSearchService.cs#L131-L132

It seems like the package is correctly found on nuget.org, but for some reason the var `filterResult` is always `NuGetPackageFilterResult.NotCompatible` (line 131).

image

If I force the var `filterResult` to `NuGetPackageFilterResult.Ok` (line 132, added by me for testing), the PM lists my plugin as expected.
With this trick, I can _install_, _use_ and _uninstall_ it (I didn't test _updating_ yet):
image

However, looking in the "Installed" tab, it still says in red the plugin is `Not compatible`:
image

### PR DESCRIPTION
**Pull request by maraf**
Resolves #27.

All searching is now in form of `id: {searchTerm} tags:GitExtensions`. This pre-filters packages marked as GitExtensions, than it loads a dependency graph (one query per package) and checks if dependency on `GitExtensions.Extensibility` exists.

Searching using nuget.org search syntax (https://docs.microsoft.com/en-us/nuget/consume-packages/finding-and-choosing-packages#search-syntax) is not ideal.

**When search term is "g", it won't find "GitExtensions.SVN"; the user must type "git", or "extensions".**
This behavior can be even seen on the web:
- https://www.nuget.org/packages?q=id%3Agitextension (0 packages)
- https://www.nuget.org/packages?q=id%3Agitextensions (6 packages)

**It seems it doesn't work well on MyGet.**
- In blog post from [2013 they said it should work](https://blog.myget.org/post/2013/03/27/improved-search-syntax-in-nuget-client-and-on-your-myget-feeds.html), but from my observation it just ignores the "id part".
- Using "id:{searchTerm}" requires to match whole packageId, nothing like "starts with" or "contains".

**It doesn't work at all when source is a folder on disk.**
- Using "id:{searchTerm}" requires to match whole packageId, nothing like "starts with" or "contains".
- Tags searching doesn't work.

### Solution
- On all feeds except local folder, load packages with tags filter, then apply id filter on loaded packages.
- On local folder feed, load all packages from feed, then apply tags and id filters on loaded packages.
- All local (in-app) filters are applied as invariant-culture ignore-case.

### CODE REVIEW

**Comment by mast-eu** [comment]
Few questions / comments:

1. _How_ did you solve #27? E.g. do you append `tags: GitExtensions` to all search terms? Do you first filter for tags and afterwards for dependency on `GitExtensions.Extensibility`?

2. > When search term is "g", it won't find "GitExtensions.SVN"; the user must type "git", or "extensions".

MS is testing a new search on nuget.org these days. Might be the reason for some strange results. https://devblogs.microsoft.com/nuget/new-and-improved-nuget-search/

3. > It doesn't work at all when source is a folder on disk.

What is the use case for this? Do you use it for debugging or is the PM internally doing this at some point?

4. > It seems it doesn't work well on MyGet.

Any idea why? If not, I'm inclined to take this behaviour as a normal bug - if it affects only not preconfigured feeds (i.e. any feed except nuget.org).

5. > A solution for the #27.

Please write `Closes #27` or `Resolves #27`. This way the issue will be automatically closed when the PR is merged. More info: https://help.github.com/en/articles/closing-issues-using-keywords#about-issue-references

6. Is this PR a draft because you're still working on it?

**Comment by maraf** [comment]
Sorry for being so brief in PR description...

1. All searching is now in form of `id: {searchTerm} tags:GitExtensions`. This pre-filters packages marked as GitExtensions, than it loads a dependency graph (one query per package) and checks if dependency on `GitExtensions.Extensibility` exists.

2. Not sure, it seems like they split package id by pascal-case and dots and that it needs to match whole word (just observation).
3. A folder on disk is a valid NuGet feed. I'm using it during debugging.
4. Today I'm going to investigate, don't have any clues yet.
5. Ook.
6. It's a draft because it only works on nuget.org, today I'm going to give it a bit more time.

**Comment by mast-eu** [comment]
> it seems like they split package id by pascal-case and dots and that it needs to match whole word (just observation).

I made the same observation. Searching for `GitExtensions.SVN` or `id:GitExtensions.SVN` or `id:"GitExtensions.SVN"` always gives results similar to `Git OR Extensions OR SVN`. Resulting in >10000 hits, although there is a package with the very same ID. Pretty useless for our case.
Anyhow, searching for `id:{searchTerm} tags:GitExtensions` seems to entirely avoid the flood of not matching packages :+1:

**Comment by maraf** [comment]
@mast-eu Last thing I'm considering is skipping id search on feeds and filter packages in PM. It's quite dirty, but until we have hundreds of plugins, it will work. Thoughts?

**Comment by mast-eu** [comment]
Getting all packages with `tags:GitExtensions` from all configured feeds directly on startup? IMHO that's ok.

**Comment by RussKie** [comment]
👍

**Comment by maraf** [comment]
I still need to fix two tests...

**Comment by maraf** [comment]
I have removed NuGet.Client update, because the never version doesn't support reordering of feeds, which we currently use. We can do the update later...

File: src/PackageManager.UI/Args.cs [hunk 1]
```diff
@@ -131,6 +135,9 @@ public override string ToString()
                 result.Append(String.Join(",", Dependencies.Select(d => d.Id + (d.Version != null ? "-v" + d.Version : ""))));
             }
 
+            if (!String.IsNullOrEmpty(Tags))
```
Comment (RussKie) [inline src/PackageManager.UI/Args.cs]: General guidelines recommend to use type aliases, i.e. `string` not `String`

File: src/PackageManager/ViewModels/PackageSourceSelectorViewModel.cs [hunk 2]
```diff
@@ -48,10 +48,13 @@ public string SelectedName
...
-                    if (value == null)
-                        service.MarkAsPrimary(null);
-                    else
-                        service.MarkAsPrimary(service.All.FirstOrDefault(s => s.Name == value));
+                    if (!isServiceUpdating)
+                    {
+                        if (value == null)
+                            service.MarkAsPrimary(null);
+                        else
+                            service.MarkAsPrimary(service.All.FirstOrDefault(s => s.Name == value));
```
Comment (RussKie) [inline src/PackageManager/ViewModels/PackageSourceSelectorViewModel.cs]: Is case sensitive comparison required? Should we perform case insensitive comparison instead?

File: src/PackageManager/ViewModels/PackageSourceSelectorViewModel.cs [hunk 3]
```diff
@@ -68,32 +71,43 @@ public PackageSourceSelectorViewModel(IPackageSourceCollection service)
...
-                SelectedName = SourceNames.FirstOrDefault(s => s == service.Primary.Name);
-            else
-                SelectedName = SourceNames.FirstOrDefault();
+                foreach (IPackageSource source in service.All)
+                    Add(source.Name);
+
+                if (isSelectedNameContained)
+                    SelectedName = selectedName;
+                else if (service.Primary != null)
+                    SelectedName = SourceNames.FirstOrDefault(s => s == service.Primary.Name);
```
Comment (RussKie) [inline src/PackageManager/ViewModels/PackageSourceSelectorViewModel.cs]: ditto

File: src/PackageManager.NuGet/Services/INuGetPackageFilter.cs [hunk 4]
```diff
@@ -3,12 +3,13 @@
...
 using System.Text;
+using System.Threading;
 using System.Threading.Tasks;
 
 namespace PackageManager.Services
 {
     public interface INuGetPackageFilter
     {
-        NuGetPackageFilterResult IsPassed(IPackageSearchMetadata package);
+        Task<NuGetPackageFilterResult> IsPassedAsync(SourceRepository repository, IPackageSearchMetadata package, CancellationToken cancellationToken);
```
[the remark this answers is not in the document]
Comment (maraf) [reply src/PackageManager.NuGet/Services/INuGetPackageFilter.cs]: Any suggestions? Does `GetFilterResultAsync` sounds better?

**Comment by RussKie** [comment]
FilterAsync?

File: src/PackageManager.NuGet/Services/NuGetSearchTerm.cs [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+{
+    public class NuGetSearchTerm : ICloneable<NuGetSearchTerm>
+    {
+        public List<string> Id { get; } = new List<string>();
+        public List<string> Version { get; } = new List<string>();
+        public List<string> Title { get; } = new List<string>();
+        public List<string> Tags { get; } = new List<string>();
+        public List<string> Description { get; } = new List<string>();
+        public List<string> Summary { get; } = new List<string>();
+        public List<string> Owner { get; } = new List<string>();
```
[the remark this answers is not in the document]
Comment (maraf) [reply src/PackageManager.NuGet/Services/NuGetSearchTerm.cs]: I'm sorry, but I don't get the point. There are supported filter parameters. As searching supports multiplicity for earch parameter, they are all defined as lists. The class is designed as mutable, as implementations of `INuGetPackageFilter` may add or remove values.

File: src/PackageManager/ViewModels/PackageSourceSelectorViewModel.cs (hunk 2 above)
Comment (maraf) [reply src/PackageManager/ViewModels/PackageSourceSelectorViewModel.cs]: Yes, it should be case-insensitive. Also all comparisons of package ids.

File: src/PackageManager.NuGet/Services/NuGetSearchTerm.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (mast-eu) [reply src/PackageManager.NuGet/Services/NuGetSearchTerm.cs]: @maraf is right. In the current design these are all supposed to be public.

**Comment by mast-eu** [comment]
Thank you

**Pull request merged**